rules/aip0127/http_template_pattern.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aip0127
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/googleapis/api-linter/lint"
+	"github.com/googleapis/api-linter/locations"
+	"github.com/googleapis/api-linter/rules/internal/utils"
+	"github.com/jhump/protoreflect/desc"
+)
+
+var (
+	resourcePatternSegment     = `[^/]+`
+	resourcePatternAnySegments = fmt.Sprintf("((%s/)*%s)?", resourcePatternSegment, resourcePatternSegment)
+	pathTemplateToRegex        = strings.NewReplacer("**", resourcePatternAnySegments, "*", resourcePatternSegment)
+)
+
+type resourceReference struct {
+	// The path of the field with the `google.api.resource_reference`. This is
+	// provided as a variable in the HTTPRule.
+	fieldPath string
+	// A template that the resource's pattern string must adhere to. This is
+	// provided by the variable's template in the HTTPRule.
+	pathTemplate string
+	// The name of the resource type. This is used to look up the resource
+	// message.
+	resourceRefName string
+}
+
+// Returns a list of resourceReferences for each variable in all the method's
+// HTTPRule's.
+func resourceRefsForMethod(m *desc.MethodDescriptor) []resourceReference {
```

Comment:
Nit: Maybe rename this to `methodResourceReferences`, I like a more noun-oriented naming when retrieving information and imperative style naming when doing an action e.g. `checkXMatchesY`.

Reply:
Done